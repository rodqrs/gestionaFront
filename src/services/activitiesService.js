import {
  API_URL,
  ACTIVITIES_PATH,
  ACTIVITIES_MANAGEMENT_PATH,
} from "../utils/const";

const activityUrl = API_URL + ACTIVITIES_PATH;
const urlActivitiesManagement = API_URL + ACTIVITIES_MANAGEMENT_PATH;

export async function getActivities() {
  try {
    const activitiesRes = await fetch(activityUrl, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });

    if (!activitiesRes.ok) throw new Error("Error fetching activities");

    const activitiesData = await activitiesRes.json();

    const newActivities = activitiesData.activities.map((activity) => {
      const {
        id: idActivity,
        nombre: name,
        descripcion: description,
        id_categoria: idCategory,
        category,
      } = activity;

      return {
        idActivity,
        name,
        description,
        idCategory,
        category: category.nombre,
        categoryDescription: category.descripcion,
      };
    });

    return newActivities;
  } catch (error) {
    console.error("There was an error retrieving the activities.", error);
  }
}

export async function getActivitiesManagement(seasonId) {
  if (!seasonId) throw new Error("Season id is required");
  if (typeof seasonId !== "string")
    throw new Error("Season id must be a string");
  const url = `${API_URL}/seasons/${seasonId}${ACTIVITIES_MANAGEMENT_PATH}`;
  try {
    const activitiesRes = await fetch(url, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });

    if (!activitiesRes.ok)
      throw new Error("Error fetching activities management");

    const activitiesData = await activitiesRes.json();

    const activitiesManagement = activitiesData.map((activity) => {
      const {
        id: idManagement,
        id_temporada: idSeason,
        id_actividad: idActivity,
        costo: cost,
        gasto_insumo_id: idSupplyExpense,
        fecha: date,
      } = activity;
      return {
        idManagement,
        idSeason,
        idActivity,
        cost,
        idSupplyExpense,
        date,
      };
    });

    return activitiesManagement;
  } catch (error) {
    console.error(
      "There was an error retrieving the activities management.",
      error
    );
  }
}

export async function createActivitiesManagement(data) {
  if (!data) throw new Error("Activities management data is required");
  if (typeof data !== "object")
    throw new Error("Activities management data must be an object");

  try {
    const activityRes = await fetch(urlActivitiesManagement, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    });

    if (!activityRes.ok) throw new Error("Error creating activity management");

    const activityData = await activityRes.json();

    const {
      id: idManagement,
      id_temporada: idSeason,
      id_actividad: idActivity,
      costo: cost,
      gasto_insumo_id: idSupplyExpense,
      fecha: date,
    } = activityData;

    return {
      idManagement,
      idSeason,
      idActivity,
      cost,
      idSupplyExpense,
      date,
    };
  } catch (error) {
    console.error("Failed to create the activity.", error.message);
  }
}

export async function deleteActivityManagement(idActivity) {
  if (!idActivity) throw new Error("Id activity management is requiered");
  if (typeof idActivity !== "string")
    throw new Error("Activity management id must be a string");

  const url = `${urlActivitiesManagement}/${idActivity}`;

  try {
    const response = await fetch(url, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
    });
    if (!response.ok) throw new Error("Error deleting activity management");

    const responsData = await response.json();

    const { id: idManagement } = responsData;
    return idManagement;
  } catch (error) {
    console.error("Failed to delete the activity", error);
  }
}

export async function updateActivityManagement(idActivity, data) {
  if (!data) throw new Error("Updated activity management data is requiered");
  if (typeof data !== "object")
    throw new Error("Updated activity management data must be an object");
  if (!idActivity)
    throw new Error("Updated activity management id is requiered");
  if (typeof idActivity !== "string")
    throw new Error("Updated activity management id must be a string");

  const url = `${urlActivitiesManagement}/${idActivity}`;

  try {
    const updatedActivityRes = await fetch(url, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    });

    if (!updatedActivityRes.ok)
      throw new Error("Error updating activity management");

    const updatedActivity = await updatedActivityRes.json();
    const {
      id: idManagement,
      id_temporada: idSeason,
      id_actividad: idActivity,
      costo: cost,
      gasto_insumo_id: idSupplyExpense,
      fecha: date,
    } = updatedActivity;

    return {
      idManagement,
      idSeason,
      idActivity,
      cost,
      idSupplyExpense,
      date,
    };
  } catch (error) {
    console.error("Failed to update the activity", error);
  }
}
