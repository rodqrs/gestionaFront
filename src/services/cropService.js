import { API_URL, CROPS_PATH, USER_SESSION } from "../utils/const";
import getUserData from "../utils/getUserData";

export async function fetchCrops() {
  const sessionUser = sessionStorage.getItem("user_data");

  if (!sessionUser) {
    console.log("Invalid project ID"); // volver al user a la vista de projects
  }

  const { projectId } = JSON.parse(sessionUser);
  const url = `${API_URL}${CROPS_PATH}/project/${projectId}`;

  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  const { crops } = await response.json();

  const newData = crops.reduce((acumulador, crop, index) => {
    const { nombre, tipo_siembra, fecha_inicio, area_terreno, unit } = crop;

    const newCrop = {
      id: index + 1,
      nombre,
      tipo_siembra,
      fecha_inicio,
      area_terreno,
      unidad: unit.unidad,
      descripcion: unit.nombre,
    };
    acumulador.push(newCrop);
    return acumulador;
  }, []);

  return newData;
}

export async function createCrop(data) {
  const sessionUser = sessionStorage.getItem("user_data");

  if (!sessionUser) throw new Error("Invalid session");

  const { projectId } = JSON.parse(sessionUser);

  if (!projectId) throw new Error("Invalid Project ID");
  if (!data || typeof data !== "object") throw new Error("Data is required");

  data["proyecto_id"] = projectId;
  data["area_terreno"] = Number(data.area_terreno);

  const url = API_URL + CROPS_PATH;

  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) throw new Error("Error creating Crop at service");

  const { crop } = await response.json();
  const { id, nombre, tipo_siembra, fecha_inicio, area_terreno, unit } = crop;

  return {
    idCrop: id,
    nombre,
    tipo_siembra,
    fecha_inicio,
    area_terreno,
    unidad: unit.unidad,
    descripcion: unit.nombre,
  };
}

export async function deleteCropByName(nameCrop) {
  if (!nameCrop) throw new Error("ID Crop is required");
  if (typeof nameCrop !== "string") throw new Error("ID Crop must be a string");

  const queryUrl = API_URL + "/query";

  const query = {
    key: "nombre",
    value: nameCrop,
    returnValue: "id",
    table: "cultivo",
  };

  const queryRes = await fetch(queryUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(query),
  });

  if (!queryRes.ok) throw new Error("Error Fetching Id");

  const queryData = await queryRes.json();

  const idCrop = queryData.data.id;

  const url = `${API_URL}${CROPS_PATH}/${idCrop}`;

  const response = await fetch(url, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new Error("Error Deleting crop");

  return {
    success: true,
  };
}

export async function updateCrop({ oldData, newData }) {
  if (!oldData) throw new Error("Old data object is required");
  if (typeof oldData !== "object")
    throw new Error("Old data must be an object");

  if (!newData) throw new Error("New data object is required");
  if (typeof newData !== "object")
    throw new Error("New data must be an object");

  let userData;
  try {
    userData = getUserData();
  } catch (error) {
    console.error("User sesion is invalid", error);
    return;
  }

  const { projectId } = userData;
  const project = userData.projectsByUser.find((proj) => proj.id === projectId);
  let crops = project.crops;

  const modifiedCrop = crops.find((crop) => crop.nombre === oldData.nombre);

  const cropUrl = `${API_URL}${CROPS_PATH}/${modifiedCrop.id}`;

  const { id: _, ...data } = newData;
  data["proyecto_id"] = projectId;
  data.area_terreno = Number(data.area_terreno);

  const updateResponse = await fetch(cropUrl, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });

  if (!updateResponse.ok) throw new Error("Error updating crop");

  const { crop: newCrop } = await updateResponse.json();

  const { id, nombre, tipo_siembra, fecha_inicio, area_terreno, unit } =
    newCrop;

  modifiedCrop.nombre = nombre;

  crops = crops.map((crop) =>
    crop.id === modifiedCrop.id ? modifiedCrop : crop
  );
  project.crops = crops;
  userData.projectsByUser = userData.projectsByUser.map((proj) =>
    proj.id === project.id ? project : proj
  );

  sessionStorage.setItem(USER_SESSION, JSON.stringify(userData));

  return {
    idCrop: id,
    nombre,
    tipo_siembra,
    fecha_inicio,
    area_terreno,
    unidad: unit.unidad,
    descripcion: unit.nombre,
  };
}
