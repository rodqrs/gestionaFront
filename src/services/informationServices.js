import { API_URL } from "../utils/const";

export async function getInformation() {
  const urls = {
    expenses: `${API_URL}/expenses`,
    sales: `${API_URL}/sales`,
    production: `${API_URL}/product`,
    supplies: `${API_URL}/supplies`,
    activitiesManagement: `${API_URL}/activities-management`,
    units: `${API_URL}/units`,
    activities: `${API_URL}/activities`,
    projects: `${API_URL}/projects`,
    crops: `${API_URL}/crops`,
    seasons: `${API_URL}/seasons`,
  };

  try {
    const responses = await Promise.all(
      Object.values(urls).map((url) =>
        fetch(url, {
          method: "GET",
          headers: { "Content-Type": "application/json" },
        }).then(response => {
          if (!response.ok) {
            console.error(`Error fetching data from ${url}:`, response.statusText);
            return null;
          }
          return response.json();
        }).catch(error => {
          console.error(`Error fetching data from ${url}:`, error);
          return null;
        })
      )
    );

    const [
      expensesData,
      salesData,
      productionData,
      suppliesData,
      activitiesManagementData,
      unitData,
      activitiesData,
      projectsData,
      cropsData,
      seasonsData,
    ] = responses;

    const expenses = expensesData?.map((element, index) => {
      const units = unitData?.units?.find((unit) => unit.id === element.id_unidad_medida);
      const supplies = suppliesData?.data?.find((supply) => supply.id === element.id_insumo);

      return {
        id: index + 1,
        suppliesName: supplies?.nombre || "No disponible",
        usedQuantity: element.cantidad_usada,
        totalPrice: element.precio_total,
        unit: units?.nombre || "No disponible",
        idSeason: element.id_temporada,
      };
    }) || [];

    const sales = salesData?.sale?.map((element, index) => {
      const units = unitData?.units?.find((unit) => unit.id === element.id_unidad_medida);

      return {
        id: index + 1,
        soldQuantity: element.cantidad_vendida || 0,
        unitPrice: element.precio_unitario || 0,
        totalPrice: element.precio_total || 0,
        saleDate: element.fecha_venta || "",
        observations: element.observaciones || "Sin observaciones",
        unit: units?.nombre || "No disponible",
        idSeason: element.id_temporada,
      };
    }) || [];

    const production = productionData?.data?.map((element, index) => {
      const units = unitData?.units?.find((unit) => unit.id === element.id_unidad_medida);

      return {
        id: index + 1,
        name: element.nombre,
        harvestedQuantity: element.cantidad_recolectada,
        harvestDate: element.fecha_recoleccion,
        unit: units?.nombre || "No disponible",
        idSeason: element.id_temporada,
      };
    }) || [];

    const activityManagements = activitiesManagementData?.map((element, index) => {
      const activity = activitiesData?.activities?.find((activity) => activity.id === element.id_actividad);
      const expenses = expensesData?.find((expense) => expense.id === element.gasto_insumo_id);
      const supplies = suppliesData?.data?.find((supply) => supply.id === expenses?.id_insumo);

      return {
        id: index + 1,
        activityName: activity?.nombre || "No disponible",
        suppliesName: supplies?.nombre || "No disponible",
        usedQuantity: expenses?.cantidad_usada || 0,
        suppliesCost: expenses?.precio_total || "No disponible",
        totalCost: element.costo,
        idSeason: element.id_temporada,
      };
    }) || [];

    return {
      expenses,
      sales,
      production,
      activityManagements,
      projects: projectsData || [],
      crops: cropsData?.crops || [],
      seasons: seasonsData || [],
    };
  } catch (error) {
    console.error("Error fetching data:", error);
    return {
      expenses: [],
      sales: [],
      production: [],
      activityManagements: [],
    };
  }
}