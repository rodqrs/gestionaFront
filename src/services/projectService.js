import {
  API_URL,
  PROJECTS_PATH,
  USERS_HAS_PATH,
  USERS_PATH,
  ROLES_PATH,
  CROPS_PATH,
  SEASONS_PATH,
} from "../utils/const";
import { isUUIDv4 } from "../utils/validations";

const url = API_URL + PROJECTS_PATH;
const roleUrl = API_URL + ROLES_PATH;
const userHasUrl = API_URL + USERS_HAS_PATH;

export async function fetchProjects() {
  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  const data = await response.json();

  return data;
}

export async function createProject(project, userId) {
  if (!project || typeof project !== "object")
    throw new Error("The project must be an object");
  if (!userId || typeof userId !== "string")
    throw new Error("The project must be an string");
  if (!isUUIDv4(userId)) throw new Error("The userId is not a valid UUIDv4");

  const fetchCreateProject = fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(project),
  });

  const fetchGetRoles = fetch(roleUrl, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  const [createProjectRes, rolesRes] = await Promise.all([
    fetchCreateProject,
    fetchGetRoles,
  ]);
  if (!createProjectRes.ok) throw new Error("Error al crear el proyecto");
  if (!rolesRes.ok) throw new Error("Error al obtener los roles");

  const [projectData, rolesData] = await Promise.all([
    createProjectRes.json(),
    rolesRes.json(),
  ]);

  const adminRoleId = rolesData.find(
    (role) => role.nombre === "Administrador"
  ).id;

  const userHasData = {
    usuario_id: userId,
    proyecto_id: projectData.id,
    id_rol: adminRoleId,
  };

  const userHasProjectRes = await fetch(userHasUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(userHasData),
  });

  if (!userHasProjectRes.ok)
    throw new Error("Error al asignar el proyecto al usuario");

  const userHasProjectData = await userHasProjectRes.json();

  return projectData;
}

export async function getProjectsByUserId(userId) {
  if (!userId) throw new Error("The userId is required");
  if (typeof userId !== "string")
    throw new Error("The userId must be a string");
  if (!isUUIDv4(userId)) throw new Error("The userId is not a valid UUIDv4");

  const urls = {
    userProjectsUrl: `${API_URL}${USERS_PATH}/${userId}${USERS_HAS_PATH}`,
    projectUrl: url,
  };

  const [userProjectsRes, projectsRes] = await Promise.all(
    Object.values(urls).map((url) =>
      fetch(url, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
      })
    )
  );

  if (![userProjectsRes, projectsRes].every((res) => res.ok)) {
    throw new Error("Error al obtener los proyectos");
  }

  const [userProjectsData, projectsData] = await Promise.all([
    userProjectsRes.json(),
    projectsRes.json(),
  ]);

  const idsUserProjects = userProjectsData.data.map(
    (project) => project.proyecto_id
  );

  const userProjects = projectsData.filter((project) =>
    idsUserProjects.includes(project.id)
  );

  return userProjects;
}

//FUNCION PARA TRAER SEASONS POR CROP ID
async function getSeasonsByCrop(id) {
  const response = await fetch(API_URL+CROPS_PATH+"/"+id+SEASONS_PATH,{
    method: "GET",
    headers: {
      "Content-Type": "application/json"
    }
  })
  const data = await response.json()

  const newData = data.reduce((acumulador, season) => {
    const {id, nombre} = season;

    const newSeason = {
      id,
      nombre
    };
    acumulador.push(newSeason);
    return acumulador;
  }, []);

  return newData
}

export async function updateSessionProjects(projects) {
  if (!projects) throw new Error("Projects is required");
  if (!(projects instanceof Array))
    throw new Error("The object must be an array of projects.");

  const dataSession = sessionStorage.getItem("user_data");

  if (!dataSession) throw new Error("The User session is invalid");

  const dataUser = JSON.parse(dataSession);

  const url = API_URL + CROPS_PATH;

  const projectsData = await Promise.all(
    projects.map(async (element) => {
      const { id, nombre } = element;
      const cropUrl = `${url}/project/${id}`;
      const res = await fetch(cropUrl, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
      });
      const cropData = await res.json();
      const crops = await Promise.all(
      cropData.crops.map(async(crop) => {
        return {
          id: crop.id,
          nombre: crop.nombre,
          seasons: await getSeasonsByCrop(crop.id),
        };
      }));

      return {
        id,
        nombre,
        crops,
      };
    })
  );

  const newUserData = {
    ...dataUser,
    projectsByUser: projectsData,
  };
  sessionStorage.setItem("user_data", JSON.stringify(newUserData));
}
