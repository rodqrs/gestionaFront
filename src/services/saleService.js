import { API_URL, SALES_PATH, SEASONS_PATH, UNITS_PATH } from "../utils/const";

const salesUrl = API_URL + SALES_PATH;
const seasonsUrl = API_URL + SEASONS_PATH;
const unitsUrl = API_URL + UNITS_PATH;
const urls = [salesUrl, seasonsUrl, unitsUrl];

export async function createSale(sale) {
  try {
    const response = await fetch(salesUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(sale),
    });
    if (!response.ok) throw new Error("Error creating sale");
    const data = await response.json();

    if (!data.success) throw new Error("Error creating sale");

    const [seasonsRes, unitsRes] = await Promise.all(
      [seasonsUrl, unitsUrl].map((url) =>
        fetch(url, {
          method: "GET",
          headers: { "Content-Type": "application/json" },
        })
      )
    );
    if (![seasonsRes, unitsRes].every((res) => res.ok)) {
      throw new Error("Una o mas repuestas de la API no fueron exitosas");
    }
    const [seasonsData, unitsData] = await Promise.all([
      seasonsRes.json(),
      unitsRes.json(),
    ]);

    const {
      cantidad_vendida: cantidadVendida,
      fecha_venta: fechaVenta,
      id,
      id_temporada: idTemporada,
      id_unidad_medida: idUnidadMedida,
      observaciones,
      precio_total: precioTotal,
      precio_unitario: precioUnitario,
    } = data.sale;

    const season = seasonsData.find((season) => season.id === idTemporada);
    const unit = unitsData.units.find((unit) => unit.id === idUnidadMedida);

    const newSale = {
      temporada: season.nombre,
      fechaVenta,
      cantidadVendida,
      unidadMedida: unit.nombre,
      precioUnitario,
      precioTotal,
      observaciones,
      idSale: id,
      idTemporada,
    };
    return newSale;
  } catch (error) {
    console.error("Error creating sale: ", error);
  }
}

export async function getSalesData() {
  try {
    const [salesRes, seasonsRes, unitsRes] = await Promise.all(
      urls.map((url) =>
        fetch(url, {
          method: "GET",
          headers: { "Content-Type": "application/json" },
        })
      )
    );
    if (![salesRes, seasonsRes, unitsRes].every((res) => res.ok)) {
      throw new Error("Una o mas repuestas de la API no fueron exitosas");
    }
    const [salesData, seasonsData, unitsData] = await Promise.all([
      salesRes.json(),
      seasonsRes.json(),
      unitsRes.json(),
    ]);
    const data = salesData.sale.map((sale, index) => {
      const {
        cantidad_vendida: cantidadVendida,
        fecha_venta: fechaVenta,
        id,
        id_temporada: idTemporada,
        id_unidad_medida: idUnidadMedida,
        observaciones,
        precio_total: precioTotal,
        precio_unitario: precioUnitario,
      } = sale;
      const season = seasonsData.find((season) => season.id === idTemporada);
      const unit = unitsData.units.find((unit) => unit.id === idUnidadMedida);

      return {
        id: index + 1,
        temporada: season.nombre,
        fechaVenta,
        cantidadVendida,
        unidadMedida: unit.nombre,
        precioUnitario,
        precioTotal,
        observaciones,
        idSale: id,
        idTemporada,
      };
    });
    return data;
  } catch (error) {
    console.error("error fetching data", error);
    return [];
  }
}

export async function getSaleByIdSeason(seasonId) {
  if (!seasonId) throw new Error("Season id is required");
  if (typeof seasonId !== "string")
    throw new Error("Season id must be a String");
  const urlSale = `${seasonsUrl}/${seasonId}${SALES_PATH}`;
  const urls = [urlSale, seasonsUrl, unitsUrl];
  try {
    const [salesRes, seasonsRes, unitsRes] = await Promise.all(
      urls.map((url) =>
        fetch(url, {
          method: "GET",
          headers: { "Content-Type": "application/json" },
        })
      )
    );
    if (![salesRes, seasonsRes, unitsRes].every((res) => res.ok)) {
      throw new Error("Una o mas repuestas de la API no fueron exitosas");
    }
    const [salesData, seasonsData, unitsData] = await Promise.all([
      salesRes.json(),
      seasonsRes.json(),
      unitsRes.json(),
    ]);
    const data = salesData.sales.map((sale, index) => {
      const {
        cantidad_vendida: cantidadVendida,
        fecha_venta: fechaVenta,
        id,
        id_temporada: idTemporada,
        id_unidad_medida: idUnidadMedida,
        observaciones,
        precio_total: precioTotal,
        precio_unitario: precioUnitario,
      } = sale;
      const season = seasonsData.find((season) => season.id === idTemporada);
      const unit = unitsData.units.find((unit) => unit.id === idUnidadMedida);

      return {
        id: index + 1,
        temporada: season.nombre,
        fechaVenta,
        cantidadVendida,
        unidadMedida: unit.nombre,
        precioUnitario,
        precioTotal,
        observaciones,
        idSale: id,
        idTemporada,
      };
    });
    return data;
  } catch (error) {
    console.error("error fetching data", error);
    return [];
  }
}

export async function deleteSale(saleId) {
  if (!saleId) throw new Error("Sale id is required");
  if (typeof saleId !== "string") throw new Error("Sale id must be a string");

  const url = `${salesUrl}/${saleId}`;

  try {
    const saleRes = await fetch(url, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
    });

    if (!saleRes.ok) throw new Error("Error deleting sale");

    const saleData = await saleRes.json();
    console.log("Delete Successful");
    return saleData.sale;
  } catch (error) {
    console.error("Error deleting sale", error);
  }
}

export async function updateSale({ saleId, newSale }) {
  if (!saleId) throw new Error("Sale Id is required");
  if (typeof saleId !== "string") throw new Error("Sale id must be a string");
  if (!newSale) throw new Error("sale updated object is required");
  if (typeof newSale !== "object")
    throw new Error("Sale updated object must be an Object");

  const url = `${salesUrl}/${saleId}`;

  try {
    const saleRes = await fetch(url, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(newSale),
    });

    if (!saleRes.ok) throw new Error("Error updating sale");

    const saleData = await saleRes.json();

    const [seasonsRes, unitsRes] = await Promise.all(
      [seasonsUrl, unitsUrl].map((url) =>
        fetch(url, {
          method: "GET",
          headers: { "Content-Type": "application/json" },
        })
      )
    );
    if (![seasonsRes, unitsRes].every((res) => res.ok)) {
      throw new Error("Una o mas repuestas de la API no fueron exitosas");
    }
    const [seasonsData, unitsData] = await Promise.all([
      seasonsRes.json(),
      unitsRes.json(),
    ]);

    const {
      cantidad_vendida: cantidadVendida,
      fecha_venta: fechaVenta,
      id,
      id_temporada: idTemporada,
      id_unidad_medida: idUnidadMedida,
      observaciones,
      precio_total: precioTotal,
      precio_unitario: precioUnitario,
    } = saleData.sale;

    const season = seasonsData.find((season) => season.id === idTemporada);
    const unit = unitsData.units.find((unit) => unit.id === idUnidadMedida);

    const newData = {
      temporada: season.nombre,
      fechaVenta,
      cantidadVendida,
      unidadMedida: unit.nombre,
      precioUnitario,
      precioTotal,
      observaciones,
      idSale: id,
      idTemporada,
    };

    console.log("Update successful");

    return newData;
  } catch (error) {
    console.error("Error updating sale", error);
  }
}
