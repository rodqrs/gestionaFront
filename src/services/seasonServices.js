import { API_URL } from "../utils/const"
const url = `${API_URL}/seasons`;
const urlCrops = `${API_URL}/crops`;
const urlNews = `${API_URL}/news`;


//GET ALL SEASON
export async function fetchSeasons() {
  const response = await fetch(url,{
    method: "GET", 
    headers: {
      "Content-Type": "application/json"
    }
  })
  const data = await response.json()
  return data
}

//GET ALL BY ID CROP
export async function fetchCropSeasons(id) {
  const response = await fetch(urlCrops+"/"+id+"/seasons",{
    method: "GET",
    headers: {
      "Content-Type": "application/json"
    }
  })
  const data = await response.json()
  // return data

  const newData = data.reduce((acumulador, season, index) => {
    const { nombre, duracion, fecha_inicio, fecha_fin,id,id_cultivo} = season;

    const newSeason = {
      id: index + 1,
      nombre,
      duracion,
      fecha_inicio,
      fecha_fin,
      id_season:id,
      id_cultivo,
    };
    acumulador.push(newSeason);
    return acumulador;
  }, []);

  return newData
}

//POST SEASON
export async function createSeason(season) {
  if (!season || typeof season !== "object")
    throw new Error("The SEASON must be an object");
  const response = await fetch(url,{
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(season)
  })
  const data = await response.json()
  return data
}

//PUT SEASON
export async function editSeason(id,season) {
  if(!id || typeof id !== "string")
    throw new Error("The ID must be an string");
  if(!season || typeof season !== "object")
    throw new Error("The SEASON must be an object");
  const response = await fetch(url+"/"+id,{
    method:"PUT",
    headers: {
      "Content-Type": "application/json"
    },
    body:JSON.stringify(season)
  })
  const data = response.json()
  return data
}

//DELETE SEASON
export async function deleteSeason(id){
  if(!id || typeof id !== "string")
    throw new Error("The ID must be an string");
  const response = await fetch(url+"/"+id,{
    method:"DELETE",
    headers:{
      "Content-Type": "application/json"
    }
  })
  const data = response.json()
  return data
}

//GET ALL Crops By Project
export async function fetchCrops() {
const sessionUser = sessionStorage.getItem("user_data");

  if (!sessionUser) {
    console.log("Invalid project ID"); // volver al user a la vista de projects
  }

  const { projectId } = JSON.parse(sessionUser);
  const urlC = `${urlCrops}/project/${projectId}`;

  const response = await fetch(urlC, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  const { crops } = await response.json();
  return crops
}
//GET ALL News
export async function fetchNews() {
  const response = await fetch(urlNews,{
    method: "GET",
    headers: {
      "Content-Type": "application/json"
    }
  })
  const data = await response.json()
  return data
}