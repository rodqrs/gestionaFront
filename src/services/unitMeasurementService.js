import { API_URL, UNITS_PATH } from "../utils/const";

const url = API_URL + UNITS_PATH

export async function fetchUnitMeasurement() {

  try {
    const response = await fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json"
      }
    })
    const data = await response.json()

    if(!data.success) throw new Error("Error fetching units measurement")
    
    return data.units
  } catch (error) {
    console.error("Error fetching unit measurement: ", error)
  }
}