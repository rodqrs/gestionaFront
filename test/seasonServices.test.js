// Importamos las funciones necesarias de vitest y nuestro servicio a probar
import { describe, it, expect, beforeEach, vi } from 'vitest'
import * as seasonServices from '../src/services/seasonServices'

// Grupo principal de pruebas para los servicios de temporadas
describe('Servicios de Temporadas', () => {
    // Datos de prueba comunes
    const mockData = {
        seasons: [
            { 
                id: '1', 
                nombre: 'Temporada 1', 
                duracion: '30 días',
                fecha_inicio: '2024-01-01',
                fecha_fin: '2024-01-31'
            }
        ],
        cropSeasons: [
            {
                id: 'season-1',
                nombre: 'Temporada 1',
                duracion: '30 días',
                fecha_inicio: '2024-01-01',
                fecha_fin: '2024-01-31',
                id_cultivo: 'crop-123'
            }
        ]
    }

    // Se ejecuta antes de cada prueba para limpiar los mocks anteriores
    beforeEach(() => {
        vi.clearAllMocks()
    })

    // Prueba para obtener todas las temporadas
    it('fetchSeasons debe obtener todas las temporadas', async () => {
        global.fetch = vi.fn(() => 
            Promise.resolve({
                json: () => Promise.resolve(mockData.seasons)
            })
        )

        const result = await seasonServices.fetchSeasons()
        // Crear snapshot del resultado
        expect(result).toMatchSnapshot()
    })

    // Prueba para obtener temporadas por cultivo específico
    it('fetchCropSeasons debe obtener temporadas por cultivo', async () => {
        const cropId = 'crop-123'
        global.fetch = vi.fn(() => 
            Promise.resolve({
                json: () => Promise.resolve(mockData.cropSeasons)
            })
        )

        const result = await seasonServices.fetchCropSeasons(cropId)
        // Crear snapshot de la transformación
        expect(result).toMatchSnapshot()
    })

    // Prueba para crear una nueva temporada
    it('createSeason debe crear una nueva temporada', async () => {
        const newSeason = {
            nombre: 'Nueva Temporada',
            duracion: '30 días',
            fecha_inicio: '2024-01-01',
            fecha_fin: '2024-01-31'
        }

        global.fetch = vi.fn(() => 
            Promise.resolve({
                json: () => Promise.resolve({ id: '1', ...newSeason })
            })
        )

        const result = await seasonServices.createSeason(newSeason)
        // Crear snapshot del resultado de creación
        expect(result).toMatchSnapshot()
    })

    // Prueba para actualizar una temporada existente
    it('editSeason debe actualizar una temporada existente', async () => {
        const seasonId = 'season-123'
        const updateData = { nombre: 'Temporada Actualizada' }

        global.fetch = vi.fn(() => 
            Promise.resolve({
                json: () => Promise.resolve({ id: seasonId, ...updateData })
            })
        )

        const result = await seasonServices.editSeason(seasonId, updateData)
        // Crear snapshot de la actualización
        expect(result).toMatchSnapshot()
        expect(fetch).toHaveBeenCalled()
    })

    // Prueba para eliminar una temporada
    it('deleteSeason debe eliminar una temporada', async () => {
        const seasonId = 'season-123'
        const mockResponse = { success: true }

        global.fetch = vi.fn(() => 
            Promise.resolve({
                json: () => Promise.resolve(mockResponse)
            })
        )

        const result = await seasonServices.deleteSeason(seasonId)
        // Crear snapshot de la respuesta de eliminación
        expect(result).toMatchSnapshot()
        expect(fetch).toHaveBeenCalled()
    })
})