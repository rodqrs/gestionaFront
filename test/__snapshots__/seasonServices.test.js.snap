// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Servicios de Temporadas > createSeason debe crear una nueva temporada 1`] = `
{
  "duracion": "30 días",
  "fecha_fin": "2024-01-31",
  "fecha_inicio": "2024-01-01",
  "id": "1",
  "nombre": "Nueva Temporada",
}
`;

exports[`Servicios de Temporadas > deleteSeason debe eliminar una temporada 1`] = `
{
  "success": true,
}
`;

exports[`Servicios de Temporadas > editSeason debe actualizar una temporada existente 1`] = `
{
  "id": "season-123",
  "nombre": "Temporada Actualizada",
}
`;

exports[`Servicios de Temporadas > fetchCropSeasons debe obtener temporadas por cultivo 1`] = `
[
  {
    "duracion": "30 días",
    "fecha_fin": "2024-01-31",
    "fecha_inicio": "2024-01-01",
    "id": 1,
    "id_cultivo": "crop-123",
    "id_season": "season-1",
    "nombre": "Temporada 1",
  },
]
`;

exports[`Servicios de Temporadas > fetchSeasons debe obtener todas las temporadas 1`] = `
[
  {
    "duracion": "30 días",
    "fecha_fin": "2024-01-31",
    "fecha_inicio": "2024-01-01",
    "id": "1",
    "nombre": "Temporada 1",
  },
]
`;
